feat: Sampler interface and ability to add custom samplers

selmanozleyen opened the pull request:
fixes: https://github.com/scverse/annbatch/issues/43

Not ready for review. Just wanted an initial blessing or feedback if your eye catched something.


Key points for my main design proposal for the interaction with `Loader` and `Sampler`:
- Loader always takes a `Sampler[list[slice]]`
- `ChunkAlignedSampler(Sampler[list[slice]])` expects an implementation `_iter_chunk_batches` which returns the chunk ids and at most two slices
- `ChunkAlignedSampler(Sampler[list[slice]])` is a subclass of it that already has `__iter__` implemented and will use `_iter_chunk_batches` and will convert chunk id's into slices.
- This way a child of `ChunkAlignedSampler` is always easily validated to be chunk aligned for our loader.

Why at most two slices per batch? Because one can always rearrange fragmantations so that a batch would have at most 2 unaligned chunk slices. I already have an idea for how a mask and category would look like following this logic.



Stuff I haven't paid attention to much:
- about rng generator storage. Like we shuffle chunks in sampler but do we need to use the same generator to shuffle batches themselves?
- worker logic. kind of ignored that
- same for `preload_nchunks` haven't given much thought
- shuffle default value (should it be optional like torch?)
- also the unit tests were vibe coded and not checked atm
- 

ping @ilan-gold , @felix0097 because its a draft PR

giovp commented:
this looks great @selmanozleyen  just chiming in cause I was working on something tangentially related that could land here as well: https://github.com/scverse/annbatch/pull/100 basically the idea of my PR is to implement the DistributedSampler like strategy, but for iterable loaders. I think it would be great if instead of being standalone, it lands into a `Sampler` implementation. 

More generally, in my opinion the most useful "samplers" that I would personally use straight away are: `DistributedSampler` and `WeightedSampler`.

ilan-gold commented on src/annbatch/sampler.py:
```diff
@@ -0,0 +1,288 @@
+        pass
+
+    @abstractmethod
+    def __len__(self) -> int:
```
Why wouldn't this be the number of iterations until exhaustion (so `n_obs // chunk_size`, for example?)

selmanozleyen replied:
this was left over. It should be, you are right

ilan-gold commented on src/annbatch/sampler.py:
```diff
@@ -0,0 +1,288 @@
+    _n_obs: int
+    _chunk_size: int
+    _preload_nchunks: int
+    _worker_handle: WorkerHandle
```
I think all samplers need a concept of `WorkerHandle` probably if we want all samplers to work with torch (for now, I think we want this to be the case)

ilan-gold reviewed (commented):
I think the base class makes sense.

I don't really get `_chunk_ids_to_slices` - why not just only return slices?

And why do we need to check unaligned slices at all? What is the failure case if we don't? It seems like too much.  This would simplify the class structure since I think `ChunkAlignedSampler` would not need to exist.

I'm also not clear what "chunk aligned" means - the on-disk sparse data doesn't really have a useful concept of chunking, so what is being aligned?

Nice start!

giovp commented on src/annbatch/sampler/_sampler.py:
```diff
@@ -0,0 +1,68 @@
+    @abstractmethod
+    def __len__(self) -> int:
+        """Return the total number of batches."""
+        ...
```
Yes, `__len__` is needed by the dataloader to compute the length of an epoch, and pass that info to stuff like progressbars but also to reset seed at the end the epoch. That is why iterable datasets, that don't necessarily have `__len__`, can't be use together with Samplers

selmanozleyen commented on uv.lock:
it was an overlook, it's fixed now

selmanozleyen commented:
This PR has gotten very big. So I won't include the categorical sampling and distributed sampling here. Once you approve of the `SliceSampler`,`LoaderBuilder` and `Loader` I suggest that we merge this to main or I create another branch on top of this one. Just to make the reviews easier.

ilan-gold commented on src/annbatch/sampler/_sampler.py:
```diff
@@ -0,0 +1,243 @@
+        batch_size: int,
+        chunk_size: int,
+        start_index: int,
+        end_index: int,
```
Why not just a `slice` object and we have one argument `mask: slice`?

ilan-gold commented on src/annbatch/loader.py:
```diff
@@ -48,14 +39,61 @@ class CSRDatasetElems(NamedTuple):
+        return type(self._train_datasets[0])
+
+
+class LoaderBuilder[
```
I am a little confused why this is necessary.  I think everything is as-you-go validateable which I tried to highlight here: https://github.com/scverse/annbatch/pull/101#discussion_r2640135047.

You add datasets (which themselves have constraints on keys, var shape etc.) and then add a `Sampler` that can have some knowledge of what is on the loader.  The `Sampler` can have a `validate` method that is abstract and must be implemented and that takes in the current state of the `Loader` object i.e., how many `obs` it has or something to validate its own settings (for example, an out-of-bounds `end_index` as its called here).

selmanozleyen replied:
Why not just enforcing no add_dataset calls after set_sampler for now? to simplify and speed up the PR. I think there are still cases to go through for a final `sampler.sample`. I went to some of them below and I think it shows just enforcing no add_dataset calls after set_sampler is the simplest way to avoid all these questions for now


---


Taking `n_obs` at iter-runtime by design means we don't want any validation related to the size of the dataset before iter. There is one case we might want validation but then it would end up us in ignoring the n_obs given at iter-time. For example consider the case where the user wants to add two datasets and wants to sample from the lower half first and upper half of second dataset. Looking like this:

```python
# dataset A: (20,n_dim)
# dataset B: (30,n_dim)
# and sample from SliceSampler(start_index=30,end_index=45) only
# user wants to give this custom sampler for this specific need

# Then would we be fine with this erroring
Loader(...).add_anndata(...).set_sampler(SliceSampler(start_index=30,end_index=45)).add_anndata(...)
# but not this?
Loader(...).add_anndata(...).add_anndata(...).set_sampler(SliceSampler(start_index=30,end_index=45))
```

Other things to consider: 
- Because in this example the user gives the optional end_index, the n_obs (or state) given to `sampler.sample(..)` won't be needed. The sampler will need to check if end_index is None, if so it won't ignore the state in `sampler.sample(..)`. Similarly in validate methods. This might complicate writing new Sampler classes in the future. 
- The only dynamic thing we can't to keep is related to the `n_obs`. I don't see why anything else should be given to the sampler.sample(...), `n_vars` and `dataset_type` are kind of locked with the first `add_dataset` method.
- Validation on values that aren't related to `n_obs` is simple and is enough to do them in `set_sampler` or `add_dataset` but we need special treatment for `n_obs` related parameters like `start_index, end_index`. Because the sampler class side shouldn't have to worry about trying to pass their custom parameters to  `sampler.sample(start_index,end_index)`

selmanozleyen replied:
> Loader(...).add_anndata(adata_with_25_obs).set_sampler(SliceSampler(mask=slice(30, None)))
> 
> I think we would want this to error out :)

ok that clarifies it for me. I was thinking maybe we'd want something call order invariant but that's not necessary. Let me revise in this PR, it's clear now.



> > The only dynamic thing we can't to keep is related to the n_obs
> 
> For now, that's all I can think of, but it's possible there will be other use-cases!

Me too, let's keep it with only n_obs for now

ilan-gold reviewed (commented):
I agree this is getting big so I would probably limit the refactoring to what is purely necessary and focus only on the `Sampler` and slotting it into the current implementation (i.e., what is on `main`), which I think is doable.  Separately, we can do a refactor PR.  In general I would aim to make the PR as small as possible! But great that things are taking shape now :) Good to see all tests pass, so seems like the basics are funcitonal

selmanozleyen commented on src/annbatch/sampler/_sampler.py:
```diff
@@ -0,0 +1,243 @@
+        self._batch_size = batch_size
+        self._chunk_size = chunk_size
+        self._shuffle = shuffle
+        self._preload_nchunks = preload_nchunks
```
so we assert `prefetch_nchunks * chunk_size % batch_size == 0` in when batch_size is not None? I still don't have a concrete idea how it would look when batch_size is None, like do we just raise a not implemented error?

Another thing to consider when simplifying the problem will be the incomplete slice we will have since asserting `(end_index - start_index) % chunk_size == 0` isn't  reasonable. We can make the incomplete slice the last shuffled slice but when there are workers it gets complicated. Since we don't have leftovers except last we need to put incomplete slice to the end. This means a dedicated worker will always need to load the last slice.

We can solve this by enforcing `drop_last` with workers and we just discard the incomplete slice

ilan-gold replied:
> so we assert prefetch_nchunks * chunk_size % batch_size == 0 in when batch_size is not None? I still don't have a concrete idea how it would look when batch_size is None, like do we just raise a not implemented error?

No, the `SliceSampler` would always need a `batch_size`.  I was referring specifically to `SliceSampler` with that comment.

> Another thing to consider when simplifying the problem will be the incomplete slice we will have since asserting (end_index - start_index) % chunk_size == 0 isn't reasonable.

Why would we assert this? I think this is almost never the case.

> We can make the incomplete slice the last shuffled slice but when there are workers it gets complicated. Since we don't have leftovers except last we need to put incomplete slice to the end. This means a dedicated worker will always need to load the last slice. We can solve this by enforcing drop_last with workers and we just discard the incomplete slice

Yes! I think this is what is happening on `main` right now? Is there an issue? `drop_last` on our `Loader` means that that specific object doesn't yield the last incomplete batch.  This works well when `torch.utils.DataLoader` has `batch_size=None` and if `torch.utils.DataLoader` has `batch_size=some_int`, `drop_last` shouldn't be applied to `Loader` but rather to `DataLoader` in which case torch should handle it judging by what's written in https://docs.pytorch.org/docs/stable/data.html#automatic-batching-default.

selmanozleyen replied:
Ok I wrote a script here it is: https://gist.github.com/selmanozleyen/99d669d1629a8ad02d259ac33773df7f

There is one way for it to work when torch batch_size=some_int and annbatch batch_size>1. That is to not unsqueeze in this line `torch.cat([d.unsqueeze(0) for d in data_list], dim=0)`. Then it doesn't error. But that leads to batches of size torch.batch_size * annbatch.batch_size often and sometimes something smaller than torch.batch_size * annbatch.batch_size because the drop_last is only on torch and not on the annbatch loader. Here is an example output when I remove the unsqueeze with annbatch.batch_size=3 and torch.batch_size=13.

```python
Created test data: 170 obs × 50 var
torch.Size([39, 50])
torch.Size([39, 50])
torch.Size([39, 50])
torch.Size([32, 50])
```
Even though the torch.drop_last is True. And in theory this batch doesn't even have to be the last batch

selmanozleyen replied:
> Yes! I think this is what is happening on main right now?

Turns out this is not the case. I read the source code and what it does is it checks if **torch's batch_size** is complete or not. So in the example I gave it dropped the last 21 obs.

```
Created test data: 170 obs × 50 var
torch.Size([39, 50]) # complete 3 batches from torch's POV (13,13,13) then call collate on it
torch.Size([39, 50]) # complete 3 batches from torch's POV (13,13,13) then call collate on it
torch.Size([39, 50]) # complete 3 batches from torch's POV (13,13,13) then call collate on it
torch.Size([32, 50]) # complete 3 batches from torch's POV (13,6,13) then call collate on it
```

https://github.com/pytorch/pytorch/blob/b3713dc2017e8dd6a2b090acac2447627093cf77/torch/utils/data/_utils/fetch.py#L30C1-L45C37

selmanozleyen replied:
> In general, I would just try to focus on making the test suite pass i.e., replicate the current behavior in a Sampler. We can worry about this once that's done or in a follow-up PR :)

oh didn't see your comment I was too focused on the investigation of the behavior. I agree I will try to prune and minimize this PR as much as possible and then we will see. But I see the conversations here still very useful for future PR's

ilan-gold commented on docs/api.md:
```diff
@@ -29,6 +29,17 @@
+   :toctree: generated/
+
+    sampler.Sampler
+    sampler.SliceSampler
```
I think this should be called `ChunkSampler` so the terminology is consistent i.e., it is parametrized by `chunk_size`

ilan-gold reviewed (changes requested):
Nice! Well on our way now :)

selmanozleyen commented on src/annbatch/loader.py:
```diff
@@ -561,107 +588,179 @@ def __iter__(
+
+            if get_worker_info() is not None:
+                worker_handle = WorkerHandle()
+                batch_sampler.set_worker_handle(worker_handle)
```
> Why can't this be done by the Sampler itself, maybe inside sample?

This can't be done in init because get_worker_info can be None i.e. torch workers not initialized. But doing it in sample is a very good idea!

selmanozleyen commented on src/annbatch/loader.py:
```diff
@@ -309,6 +326,29 @@ def add_dataset(self, dataset: BackingArray, obs: pd.DataFrame | None = None) ->
+            self._batch_sampler.validate(self.n_obs)
+        return self
+
+    def set_sampler(self, sampler: Sampler[list[slice]]) -> Self:
```
@ilan-gold it sounded very reasonable and I was happy that it was getting simplified but now it fails when the user wants to set something custom:

```python
n_obs = sum([d["dataset"].shape[0] for d in datas])
start_idx, end_idx = n_obs // 4, n_obs // 2
sampler = SliceSampler(
    mask=slice(start_idx, end_idx),
    batch_size=10,
    slice_size=10,
    preload_nslices=2,
)
loader = Loader(preload_to_gpu=False, to_torch=False, return_index=True, batch_sampler=sampler)
loader.add_datasets(**concat(datas)) # validate fails here in add_dataset since `n_obs` isn't yet up to date
```
Only sensible thing that comes to my mind is to not support add_dataset with Loader which is probably because I am still biased to a builder oriented strategy.

So until its clarified I will keep the cannot_set_sampler thing.

And to note this isn't some obscure use case we will want to specify category boundaries in our sampler.

ilan-gold replied:
I think there is already special casing in the code within `add_dataset` for what happens when we are adding something for the first time, so this would be nothing special too code to `add_datasets`/`add_anndatas` that check if there are no datasets, and then if not, performs validation up front.  Then we can rely on `add_dataset`.  We can create a private `_add_dataset(..., *, validate_sampler: bool)` method that does not validate and is used by `add_datasets`/`add_anndatas` on the first addition with `validate_sampler=True` so we don't have to change the public API.  Does this seem reasonable?

>  Like if a user prefers to call add_dataset consecutively instead of add_datasets they are doing it wrong for our case.

This I'm ok with.  With the way things are headed in https://github.com/scverse/annbatch/issues/94#issuecomment-3638425459, `add_dataset` is going to become an even  lower-level API anyway.

selmanozleyen commented on src/annbatch/loader.py:
```diff
@@ -561,107 +588,179 @@ def __iter__(
+        if self._to_torch:
+            data = to_torch(in_memory_data[split], self._preload_to_gpu)
+        else:
+            data = in_memory_data[split]
```
thanks

selmanozleyen commented:
Those two sound reasonable, in fact I tried to implement that but didn't had to time to make sure it worked well with the tests. Basically the same idea you got. The incomplete slice would also be randomly selected

ilan-gold commented on src/annbatch/sampler/_sampler.py:
```diff
@@ -0,0 +1,284 @@
+
+    @property
+    @abstractmethod
+    def worker_handle(self) -> WorkerHandle | None:
```
Not entirely clear how this works since `WorkerHandle` is not exported (i.e., docs).  I think there should just be a default implementation for now.  We can worry about torch worker stuff later.

ilan-gold replied:
Separate from the above, I also think this doesn't need to be public (it's only used privately?)

ilan-gold commented on src/annbatch/sampler/_sampler.py:
```diff
@@ -0,0 +1,284 @@
+        *,
+        batch_size: int,
+        slice_size: int,
+        mask: slice,
```
Let's validate the `mask`'s step size as well (I think it's called `slice.step` i.e., it should only be 1 or `None` if `None` is the default that is actually 1)

selmanozleyen replied:
good point

ilan-gold commented on src/annbatch/loader.py:
```diff
@@ -279,7 +299,12 @@ def add_dataset(self, dataset: BackingArray, obs: pd.DataFrame | None = None) ->
+        self._add_dataset(dataset, obs)
+        return self
+
+    def _add_dataset(self, dataset: BackingArray, obs: pd.DataFrame | None = None) -> Self:
```
Would call this `_add_dataset_unchecked` or something

ilan-gold commented on tests/test_sampler.py:
```diff
@@ -0,0 +1,671 @@
+        n_iters = math.ceil(n_obs / obs_per_iter)
+
+        expected_sizes_per_iter = []
+        leftover = 0
```
Is there ever leftover? I don't think so give the settings here

ilan-gold reviewed (changes requested):
Nice!

selmanozleyen commented on src/annbatch/loader.py:
```diff
@@ -111,53 +104,71 @@ class Loader[
+            self._preload_nslices = sampler_args_filtered.get("preload_nslices", self._preload_nslices)
+            self._batch_size = sampler_args_filtered.get("batch_size", self._batch_size)
+            self._shuffle = sampler_args_filtered.get("shuffle", self._shuffle)
+            self._drop_last = sampler_args_filtered.get("drop_last", self._drop_last)
```
right, no reason to store them anymore.

selmanozleyen commented on src/annbatch/sampler/_sampler.py:
```diff
@@ -0,0 +1,284 @@
+            splits = splits[:-1]
+        # only doesn't happen when preload_nslices * slice_size = batch_size and drop_last=True
+        if len(splits) > 0:
+            yield LoadRequest(slices=[slices[idx] for idx in slices_per_iter[-1]], splits=splits)
```
ok rest is perfect, I forgot about `_process_iter`, it was an artifact from when the function was more complicated.

selmanozleyen committed:
parent 627eb08d699b9cb07ab24fa67775e1e794c07245
author selmanozleyen <syozleyen@gmail.com> 1768761446 +0100
committer selmanozleyen <syozleyen@gmail.com> 1768846971 +0100

resolve conflicts with main

update tests to resolve conflict

readthedocs merge

merge changes

apply suggestions

checkout readme from main

selmanozleyen committed:
resolve conflicts with main

selmanozleyen committed:
load_obs thing was removed by auto formatting

selmanozleyen committed:
update tests to resolve conflict

selmanozleyen committed:
readthedocs merge

ilan-gold committed:
chore: clarify compatibility of `h5ad` + forward compat of old shuffled `zarr` (#114)

* chore: clarify compatibility of `h5ad` + forward compat of old shuffled `zarr`

* chore: version

* fix: docs

* clarify warning

* fix: more

* fix: `add_anndata`

* fix: `h5ad` compat

ilan-gold committed:
breaking: clarify obs handling + change output keys (#115)

* chore: clarify obs handling

* chore: clearer docs

* fix: rename

ilan-gold committed:
breaking: clarify obs handling + change output keys (#115)

* chore: clarify obs handling

* chore: clearer docs

* fix: rename

ilan-gold committed:
breaking: clarify obs handling + change output keys (#115)

* chore: clarify obs handling

* chore: clearer docs

* fix: rename

ilan-gold committed:
fix: header level (#116)

ilan-gold committed:
fix: header level (#116)

ilan-gold commented on src/annbatch/sampler/_sampler.py:
```diff
@@ -0,0 +1,242 @@
+        shuffle: bool = False,
+        preload_nchunks: int,
+        drop_last: bool = False,
+        rng: np.random.Generator | None = None,
```
```suggestion
        chunk_size: int,
        preload_nchunks: int,
        batch_size: int,
        *,
        mask: slice | None = None,
        shuffle: bool = False,
        drop_last: bool = False,
        rng: np.random.Generator | None = None,
```

ilan-gold reviewed (changes requested):
Mostly small things now, awesome!

selmanozleyen committed:
merge changes

selmanozleyen committed:
apply suggestions

selmanozleyen committed:
checkout readme from main

selmanozleyen committed:
restore from main

selmanozleyen committed:
fix: checking out: confused origin and upstream again...

selmanozleyen committed:
continuation of the upstream origin confusion fix

selmanozleyen committed:
Merge branch 'main' into feat/sampler

selmanozleyen committed:
refactor _prepare_dataset_and_obs

selmanozleyen committed:
update docstring for loadrequest

selmanozleyen committed:
separate files for samplers

selmanozleyen commented on src/annbatch/loader.py:
```diff
@@ -590,107 +601,92 @@ def __iter__(
+
         in_memory_data = None
         concatenated_obs = None
         in_memory_indices = None
```
see latest refactor, they aren't needed indeed

selmanozleyen committed:
prepare_output is no longer needed

selmanozleyen committed:
clarify docs

selmanozleyen committed:
fix overlook: already sorted batch_indices no need to resort them

selmanozleyen committed:
fix prepare_output refactor

selmanozleyen committed:
add todo

selmanozleyen committed:
rename from leftover to remainder for clarity. since there is no leftover

selmanozleyen committed:
simplify validate_sampler

selmanozleyen committed:
remove old generic params

selmanozleyen committed:
add broad typing

selmanozleyen committed:
clarify todos and add username

selmanozleyen committed:
type and modify decorator

selmanozleyen committed:
no need for lambdas in decorators

selmanozleyen committed:
make decorator compatible in multiple cases

selmanozleyen committed:
put ABC in abc folder

selmanozleyen committed:
update test with the fix

selmanozleyen committed:
qualname for fix. no sampler in public API

selmanozleyen committed:
check coverage when shuffled otherwise also check order

selmanozleyen committed:
fix to prev commit

selmanozleyen committed:
clarify doc

selmanozleyen committed:
update worker tests

selmanozleyen committed:
new * location for ChunkSampler

selmanozleyen commented on tests/test_sampler.py:
```diff
@@ -0,0 +1,238 @@
+
+    expected_start = start if start is not None else 0
+    expected_stop = stop if stop is not None else n_obs
+    assert all_indices == set(range(expected_start, expected_stop))
```
done

selmanozleyen committed:
add typing but can revert if too verbose

selmanozleyen committed:
remove unused fields. (maybe linter check can be added)

selmanozleyen committed:
remove old SO link

selmanozleyen committed:
don't put generators into np.all !!

ilan-gold commented on src/annbatch/loader.py:
```diff
@@ -68,30 +67,42 @@ class Loader[
+            - `preload_nchunks`: 32
+            - `batch_size`: 1
+            - `shuffle`: False
+            - `drop_last`: False
```
```suggestion
            If not provided, a default :class:`annbatch.ChunkSampler` will be used with the same defaults below.
```

ilan-gold commented on src/annbatch/utils.py:
```diff
@@ -15,11 +15,46 @@
+
+def validate_sampler[Self: HasBatchSampler, **Param, RetType](
+    method: Callable[Concatenate[Self, Param], RetType],
+) -> Callable[Concatenate[Self, Param], RetType]:
```
```suggestion
def validate_sampler[**Param, RetType](
    method: Callable[Concatenate[Loader, Param], RetType],
) -> Callable[Concatenate[Loader, Param], RetType]:
```
Do we even need `Self` + `Protocol`?

ilan-gold commented on src/annbatch/sampler/_chunk_sampler.py:
```diff
@@ -0,0 +1,182 @@
+        chunk_indices = np.arange(math.ceil((stop - start) / self._chunk_size))
+        if self._shuffle:
+            # TODO(selmanozleyen): maybe this should be done worker-aware way?
+            self._rng.shuffle(chunk_indices)
```
I think (can't be certain) that this is only working "by accident" in so far as the `_rng` just happens to have the same state across workers, but I assume there is a way to break that criterion (don't know enough about forking a RNG to be certain). 
```suggestion
        if self._shuffle:
            worker_handle.shuffle(chunks)
```

ilan-gold reviewed (commented):
Small things now just!

selmanozleyen committed:
apply typing and docstring suggestion

selmanozleyen committed:
change in folder structure

selmanozleyen committed:
make batch sampler getter

selmanozleyen committed:
remove empty line

selmanozleyen committed:
apply docstring suggestions for Loader args

selmanozleyen committed:
remove empty line

selmanozleyen committed:
conf.py is same as main

selmanozleyen committed:
change shuffle

selmanozleyen committed:
remove todo

selmanozleyen committed:
update to match old behaviour

ilan-gold reviewed (approved):
Ok This looks good! Great stuff - how do you want to proceed here? Do you want to try implementing a new sampler off this branch or merge this as-is?

ilan-gold reviewed (commented):
> Nicee!! I'd like to merge this as is if you don't mind. It's already big enough

Ok!

selmanozleyen committed:
Merge branch 'main' into feat/sampler

selmanozleyen committed:
put vstack inside accumulate chunks
